File: Binary_Tree/257.Binary_Tree_Paths/C++/257.Binary_Tree_Paths.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<string> binaryTreePaths(TreeNode * root) 
    {
        //====================================//
        // Input type:                        //
        //   - TreeNode * (root tree-node)    //
        // Return type:                       //
        //   - vector<string> (result vector) //
        //====================================//

        /*Initialize*/
        ///// Result vector /////
        vector<string> res_vect;

        /*Recursion-based DFS loop traversal*/
        dfsTraversal(root, to_string((root->val)), res_vect); //Recursion function call (i.e. whole root tree-node)

        return res_vect;
    }


    void dfsTraversal(TreeNode * root, string record_str, vector<string> & res_vect)
    {
        //=========================================//
        // Input type:                             //
        //   - TreeNode * (root tree-node)         //
        //   - string (record string)              //
        //   - vector<string> ref. (result vector) //
        // Return type:                            //
        //   - void (no return)                    //
        //=========================================//
        if (((root->left) == nullptr) && ((root->right) == nullptr)) //Issue, Boundary-case handling
        {
            res_vect.emplace_back(record_str); //Update, Record

            return;
        }

        /*Whole process, flow*/
        ///// Check if the current root left-half node existed or not /////
        if ((root->left) != nullptr)
        {
            dfsTraversal((root->left), (record_str + "->" + to_string(((root->left)->val))), res_vect); //Recursion function call (i.e. root left-half node)
        }

        ///// Check if the current root right-half node existed or not /////
        if ((root->right) != nullptr)
        {
            dfsTraversal((root->right), (record_str + "->" + to_string(((root->right)->val))), res_vect); //Recursion function call (i.e. root right-half node)
        }
    }
};

File: Binary_Tree/257.Binary_Tree_Paths/Python/257.Binary_Tree_Paths.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def binaryTreePaths(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: List[str]
        """
        #======================================#
        # Recursion-based DFS traversal method #
        #======================================#

        ############
        #Initialize
        ##### Result array #####
        res_arry = []

        ####################################
        #Recursion-based DFS loop traversal
        self.dfsTraversal(root, str(root.val), res_arry) #Recursion function call (i.e. whole root tree-node)

        return res_arry


    def dfsTraversal(self, root, record_str, res_arry):
        """
        :type root: Optional[TreeNode]
        :type record_str: str
        :type res_arry: List[str]
        :rtype: None, void
        """
        #========================================#
        # Recursion-based DFS traversal function #
        #========================================#
        if ((not root.left) and (not root.right)): #Issue/Boundary-case handling
            res_arry.append(record_str) #Update/Record

            return

        ##### Check if the next root left-half node existed or not ##### 
        if (root.left):
            self.dfsTraversal(root.left, (record_str + "->" + str((root.left).val)), res_arry) #Recursion function call (i.e. root left-half node)

        ##### Check if the next root right-half node existed or not #####
        if (root.right):
            self.dfsTraversal(root.right, (record_str + "->" + str((root.right).val)), res_arry) #Recursion function call (i.e. root right-half node)
